packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -77,20 +74,24 @@ export class BlobManager {
             `${BlobManager.basePath}/${response.id}`,
             this.routeContext,
             async () => this.getStorage().readBlob(response.id),
-            () => this.attachBlobCallback(response.id),
         );
 
         // Note - server will de-dup blobs, so we might get existing blobId!
         if (!this.blobIds.has(response.id)) {
-            this.pendingBlobIds.add(response.id);
+            this.pendingBlobIds.set(response.id, new Deferred<void>());
+
+            // send blob attach op and wait until we see it to return the handle
+            this.attachBlobCallback(response.id);
+            await this.pendingBlobIds.get(response.id)?.promise;
         }
 
         return handle;
     }
 
     public addBlobId(blobId: string) {
```

Comment:
Can we break it into two functions - one private addBlobCore() called from load() (that would only add to blobIds) and another - postBlobAttach() or something of similar name called from container runtime? It would be easier to follow intent. Also for the later, it should be unconditional, i.e. there has to be an entry for it in this.pendingBlobIds when it's called (if we handle concurrent uploads of same image correctly, see prior comment)